### ISSUE TITLE
[BUG] Unable to build libwebsockets

### ISSUE THREAD

**Issue by fernandoc1** [awslabs/amazon-kinesis-video-streams-webrtc-sdk-c#556]
I'm trying to compile the code for Amazon Kinesis with gcc version 4.8.3 20131202 (Hisilicon_v300) for an ARM board.

My compilation command is:

`CC=arm-hisiv300-linux-gcc CXX=arm-hisiv300-linux-g++ AR=arm-hisiv300-linux-ar cmake .. -DBUILD_OPENSSL=TRUE -DBUILD_OPENSSL_PLATFORM=linux-generic32 -DBUILD_LIBSRTP_HOST_PLATFORM=x86_64-unknown-linux-gnu -DBUILD_LIBSRTP_DESTINATION_PLATFORM=arm-unknown-linux-uclibcgnueabi`

I'm facing the following error:

```
amazon-kinesis-video-streams-webrtc-sdk-c/open-source/libwebsockets/build/src/project_libwebsockets/lib/roles/pipe/ops-pipe.c:30:2: error: implicit declaration of function ‘eventfd_read’ [-Werror=implicit-function-declaration]
  if (eventfd_read(wsi->desc.sockfd, &value) < 0)

```

**Comment by fernandoc1** [comment]
I fixed my compilation issues with the following changes:
[](https://gist.github.com/fernandoc1/e2b6821bf7431caec8bac7d8dd47921c)

**Comment by MushMal** [comment]
@fernandoc1 this is a good change I believe. Could you please submit a PR? All you need to do is rebase on the latest master branch, git checkout -b <your_branch_name> and git push. This will produce a link for the PR. Please submit it so we can integrate.

**Comment by fernandoc1** [comment]
Done:
[557](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/557)

**Comment by ycyang1229** [comment]
@fernandoc1 Hey, the platform is interesting. Does linux run on your platform or other non-linux-based RTOS?

**Comment by fernandoc1** [comment]
My platform is Linux based.
I can share my compiler toolchain for kinesis testing it's compatibility.

**Comment by MushMal** [comment]
@fernandoc1 after some thinking it might be possible that the fix you've submitted could break some customers. In order to not regress, the customers who are affected by this we recommend modifying the appropriate lib...Cmake directly in their environment. We will try to think how we can improve this custom flags passing to third party libs better.

Resolving this for now, please don't hesitate to re-open

**Comment by suggestedfixes** [comment]
Don't worry about making new changes, I assure you have broken the build multiple times :P

May I suggest to make a new CMAKE flag for this particular type of platform? (I've witnessed humongous systems working seamlessly with millions of flags)
This way we keep the old behaviors. And for customers using a similar platform, they can try out this new flag. If this new flag doesn't work, there's nothing to lose.

**Comment by MushMal** [comment]
@suggestedfixes we have considered various approaches to platforms which we don't directly support but we really need to be able to draw a line. Almost every single build issue we had had has been absolutely unrelated to the KVS SDKs but rather building or linking it's dependencies. While we still don't have a good solution, we try to at least keep the minimal build workable in the TravisCI configurations we employ in the projects. I do think having some options of passing build flags into the dependencies is a nice to have but does it warrant the complexity? Many developers working on proprietary platforms know better how to directly modify the CMAke files for their platforms anyway.

**Comment by suggestedfixes** [comment]
"I do think having some options of passing build flags into the dependencies is a nice to have but does it warrant the complexity?"
Yes, but you spend less support time on custom builds and hours of customer meetings complaining. So that you can have more time on the fun part, development.

"Many developers working on proprietary platforms know better how to directly modify the CMake files for their platforms anyway."
True, but we will still drain your time supporting custom builds. Also, there are only O(1) types of CPU out there, once they've been written, they are probably solid.

Also, I understand my arguments are flawed and naive. But I understand one of the factors is to find a balance between code complexity and the amount of support needed.

**Comment by suggestedfixes** [comment]
"https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/system-requirements.html"
Here is a list of advertised supported hardware for kvs producer.
I think it's reasonable to define bare min to be those ones for WebRTC as well.

**Comment by MushMal** [comment]
@suggestedfixes perhaps we should set the correct expectations as well. The documentation you bring in talks only about the hardware that we tried internally. This has no bearing on whether the TravisCI build system will support it in the GitHub nor whether WebRTC is supported on those. While I completely understand that it would be very useful for some customers such as yourself to have their hardware supported out-of-the-box, it will be virtually impossible to support all of them. Getting a single set of build scripts working with all of these dependencies (however minimal they might be) is a very tough challenge, especially, since most platforms would actually have their own copies of the binaries in the image. 

The core of the code itself is possible scale up and down and sideways on pretty much any hardware, any compiler and any  platform but the build system for the dependencies causes us a lot of pain overall and this is a task that has not been properly solved overall.

We should continue iterating on this subject and take each issue and discuss it. If you guys feel that a particular set of flags for a given platform for the given dependency warrants its own switch/param at the higher level build of KVS client assets then let's do it but it will have to work through the layers we have (WebRTC for example, is based on top of Common in C Producer SDK, which, in turn, is based on PIC).

Meanwhile, some of the build integration can be achieved by having a good documentation on what to change and how to add switches and params in the CMake files.

Thoughts?

**Comment by fernandoc1** [comment]
I think that putting the dependencies as a submodule of kinesis would help fix some issues, because we could apply automatic patches to those modules and then fix particular platform issues.
When I'm compiling the source, I really don't like the fact that dependencies are downloaded as part of the CMAKE configuration process.
If I need to change anything in the downloaded modules, these changes are lost. Unless I manually customize kinesis CMake files.

**Comment by fernandoc1** [comment]
I would like to share a patch I created to compile Kinesis in my platform. Is there any way to make it easy to compile in future releases?

<pre>
diff --git a/CMake/Dependencies/libopenssl-CMakeLists.txt b/CMake/Dependencies/libopenssl-CMakeLists.txt
index c5da8a57..d9cb3d4c 100644
--- a/CMake/Dependencies/libopenssl-CMakeLists.txt
+++ b/CMake/Dependencies/libopenssl-CMakeLists.txt
@@ -12,7 +12,7 @@ else()
   endif()
 
   if (DEFINED BUILD_OPENSSL_PLATFORM AND NOT BUILD_OPENSSL_PLATFORM STREQUAL OFF)
-    SET(CONFIGURE_COMMAND ${CMAKE_CURRENT_BINARY_DIR}/build/src/project_libopenssl/Configure ${OPENSSL_EXTRA} no-async --prefix=${OPEN_SRC_INSTALL_PREFIX} --openssldir=${OPEN_SRC_INSTALL_PREFIX} ${BUILD_OPENSSL_PLATFORM})
+    SET(CONFIGURE_COMMAND ${CMAKE_CURRENT_BINARY_DIR}/build/src/project_libopenssl/Configure ${OPENSSL_EXTRA} --prefix=${OPEN_SRC_INSTALL_PREFIX} --openssldir=${OPEN_SRC_INSTALL_PREFIX} ${BUILD_OPENSSL_PLATFORM})
   else()
     SET(CONFIGURE_COMMAND ${CMAKE_CURRENT_BINARY_DIR}/build/src/project_libopenssl/config ${OPENSSL_EXTRA} --prefix=${OPEN_SRC_INSTALL_PREFIX} --openssldir=${OPEN_SRC_INSTALL_PREFIX})
   endif()
diff --git a/CMake/Dependencies/libwebsockets-CMakeLists.txt b/CMake/Dependencies/libwebsockets-CMakeLists.txt
index 4f48f9ba..24e832d6 100644
--- a/CMake/Dependencies/libwebsockets-CMakeLists.txt
+++ b/CMake/Dependencies/libwebsockets-CMakeLists.txt
@@ -39,7 +39,6 @@ ExternalProject_Add(project_libwebsockets
         -DLWS_WITH_SHARED=${LWS_WITH_SHARED}
         -DLWS_STATIC_PIC=1
         -DLWS_WITH_ZLIB=0
-        -DLWS_HAVE_EVENTFD=0
         -DLWS_WITH_MBEDTLS=${LWS_WITH_MBEDTLS}
 # enable for debug output        -DCMAKE_BUILD_TYPE=DEBUG
         -DOPENSSL_ROOT_DIR=${OPENSSL_DIR}
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 301ab154..12af4f4e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -252,7 +252,6 @@ target_link_libraries(
   kvsWebrtcClient
   PRIVATE kvspicUtils
           kvspicState
-          m
           ${CMAKE_THREAD_LIBS_INIT}
           ${OPENSSL_SSL_LIBRARY}
           ${OPENSSL_CRYPTO_LIBRARY}
</pre>

[hisilicon_build.txt](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/files/5104643/hisilicon_build.txt)

**Comment by MushMal** [comment]
@fernandoc1 thanks for submitting a PR: https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/557/files

I have a concern about the addition of m library. Why is this necessary in the kvsWebrtcClient? Not sure if this is due to the PIC defining include <math.h> in the common defs? Is this due to a compile time issue of missing definition or a linker error of some sort?

**Comment by fernandoc1** [comment]
@MushMal, the problem is with "ldexp" symbol is being complained when linking "libkvsWebrtcClient.so".
If I remove the "m" flag, I get this error:
<pre>
libkvsWebrtcClient.so: undefined reference to `ldexp'
</pre>
The cause for this, I really don't know. But what I did was to just grep for this symbol in my toolchain and I found it in libm.
I wish to have a clean compilation experience. This would help me a lot in the future.

**Comment by MushMal** [comment]
Ah, that one... Thanks for pointing out. There really isn't any need to use ldexp here at all. The exponent is a positive value so it can be replaced easily.

**Comment by fernandoc1** [comment]
When what about the other lines I need?
The addition of DLWS_HAVE_EVENTFD=0 to websockets and no-async to compile openssl?
Can you help me get a better experience compiling the code?

**Comment by MushMal** [comment]
I believe no-async is the right change and I am OK. DLWS_HAVE_EVENTFD=0 I don't have much info but if it works in Travis CI runs I am good. 

Here is what I recommend. Let's land the change I just submitted #723 and in parallel, you can update your PR and remove 'm' library usage.

**Comment by fernandoc1** [comment]
I have created another merge request:
[557 - Fixed compilation issues](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/557)

**Comment by MushMal** [comment]
Resolving this issue for now. The script changes are in. We are about to land the code changes shortly which are in PR #723
